File: drills/elixir/solutions/nc96_stock_with_cooldown.exs
defmodule Solution do
  @impossible -1_000_000_000

  # Three states rather than one number: holding a share, having just sold (so
  # today is the cooldown), and free to act. Each day's states depend only on
  # yesterday's, so the whole thing is three rolling values -- and the cooldown
  # is expressed simply by "free" never reading "sold" from the same day.
  def max_profit(prices) do
    {_hold, sold, rest} =
      Enum.reduce(prices, {@impossible, @impossible, 0}, fn price, {hold, sold, rest} ->
        {max(hold, rest - price), hold + price, max(rest, sold)}
      end)

    max(max(sold, rest), 0)
  end
end

File: drills/go/solutions/nc96_stock_with_cooldown.go
package main

func maxProfit(prices []int) int {
	// Three states carried day to day: holding a share, just sold (must
	// cool down tomorrow), and free to buy.
	holding, cooling, free := -1<<31, 0, 0
	for _, price := range prices {
		holding, cooling, free = max(holding, free-price), holding+price, max(free, cooling)
	}
	return max(cooling, free)
}
